Feat/oci registry collector

robert-cronin opened the pull request:
# Description of the PR

<!-- Please include a summary of the change, including relevant motivation and context. -->

<!-- If this PR fixes an issue, please state this using "Fixes #XYZ" -->
This is my attempt at progressing #1279 cc @ridhoq 
I think this feature is still one we want to add long-term

Partial fix for #298, this should be merged first and then #2241 will fix the issue.

A couple of thoughts I've had along the way:
- It seems like a call to `/v2/_catalog` requires authentication for many registries (e.g. ghcr.io, ghr.io to name two)
- My understanding is that `regclient` silently fails on a 401 response from the registry and doesn't seek to reconcile it by attempting login with local creds even when we pass WithDockerCreds in. I'm not sure if that is still the case. Relevant issue: https://github.com/regclient/regclient/issues/748#issuecomment-2122725832
- Another useful issue/comment for context: https://github.com/regclient/regclient/issues/792#issuecomment-2329783099
- In my drafting of testing for the feature, I could only get `mcr.microsoft.com` to return a list of repos (see https://mcr.microsoft.com/v2/_catalog for this list) however the call returns around 2k+ repos with some having upwards of 100+ tags, so unless we implement some kind of concurrency, I think unit testing might be infeasible. Also, it looks like mcr doesn't yet support limits (e.g. https://mcr.microsoft.com/v2/_catalog?n=5).

Any guidance on the testing or concurrency would be appreciated!

# PR Checklist

- [x] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [x] All new changes are covered by tests
- [ ] If GraphQL schema is changed, `make generate` has been run
- [ ] If GraphQL schema is changed, GraphQL client updates/additions have been made
- [ ] If OpenAPI spec is changed, `make generate` has been run
- [ ] If ent schema is changed, `make generate` has been run
- [ ] If `collectsub` protobuf has been changed, `make proto` has been run
- [ ] All CI checks are passing (tests and formatting)
- [ ] All dependent PRs have already been merged

robert-cronin commented:
This is still a WIP, but thought I should push what I have so far. A couple more thoughts:
- I think mocking the registry endpoints is probably the most feasible way to get reliable unit testing. To this end, I have adapted a httptest handler [found in the regclient lib](https://github.com/regclient/regclient/blob/e8054262030074b96c15671e0e0d86d79233e7c7/internal/reqresp/reqresp.go#L96) that returns mock data for the catalog, manifests and tags endpoints.
- To address the concern of potentially having all repos in a registry collected serially, there are two mechanisms that might be suitable:
    - processing the tags of an image in a worker queue which I have attempted in the latest commit; and
    - passing each new repo into CollectSub via gRPC which I have yet to attempt.
Any feedback or suggestions would be greatly appreciated!
cc @ridhoq

robert-cronin commented:
I've created a basic test suite which is working for the registry collector. It is an attempt to model the [Pull Workflow](https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#pull) that starts from a call to the catalog endpoint. Below are a couple of points that might be of interest to reviewers:
- OCIRegistryCollector is essentially a wrapper over OCICollector so the collected documents will have the source labelled as OCICollector and not OCIRegistryCollector. Not sure how best to fix this without duplicating the OCICollector code but maybe it is okay.
- There is probably an easier way to test the registry collector; it was useful for my learning and I figured it might also be useful for other test suites in the future, but I am happy to find some other way to test the feature as well.

pxp928 commented:
@robert-cronin is this ready for review?

robert-cronin commented:
> @robert-cronin is this ready for review?

yes, all ready for review 👍

robert-cronin commented:
I've now addressed the TODO comments and also tried to add the registry collector to the collect sub although I'm not confident I haven't missed anything there; I think I am still confused about how exactly the CollectSources get picked up by the respective collectors.

robert-cronin commented:
just to add a bit of extra context; in my local testing I have run the collector against a local ACR registry with a mock image + referrers and that seems to be working (assuming the 404's are expected to be there, I guess they are a part of the OCI collector logging?)
Screenshot below:
![Screenshot 2024-10-22 173457](https://github.com/user-attachments/assets/f6cd4929-9476-4d4b-b195-e2912f642800)
could probably have added a known manifest type but I think it demonstrates that the registry collector is successfully indexing and passing the images onto the oci collector.

pxp928 commented:
Thanks @robert-cronin! Let me take a look first thing in the morning and I will review your PR.

robert-cronin commented:
> Thanks @robert-cronin! Let me take a look first thing in the morning and I will review your PR.

@pxp928 no problems, feel free to take your time. Thank you!

robert-cronin commented on pkg/handler/collector/oci/oci_registry_test.go:
```diff
@@ -0,0 +1,366 @@
+	}
+}
+
+type mockHandler struct {
```
Sure thing, I'll refactor it to be more usable!

robert-cronin replied:
I've added a mock registry under [internal/testing/mockRegistry/mock_registry.go](https://github.com/robert-cronin/guac/blob/9272eb5c4ae0556015201d86f90c5a19bb4b6c51/internal/testing/mockRegistry/mock_registry.go) and exposed it publicly for others to use. I've tried to make it as general and faithful as possible to the distribution spec, but there might be areas that don't fully capture the input/output behavior of a proper registry implementation. I've also added tests for the mock registry to make sure it's as reliable as possible as a replacement for an actual registry.

robert-cronin commented on cmd/guaccollect/cmd/oci.go:
```diff
@@ -46,6 +47,19 @@ type ociOptions struct {
 	publishToQueue bool
 }
 
+type ociRegistryOptions struct {
```
The OCI collector and OCI registry collector are both using the docker credentials through `regclient.WithDockerCreds`. As I understand it, that just means the creds inside `~/.docker/config.json` are used:

https://github.com/guacsec/guac/blob/0a415eb77e00dba8f3c22ab494c9fbe7e3860761/pkg/handler/collector/oci/oci.go#L533

robert-cronin replied:
So one would need to first login to the registry locally, I could be mistaken and there is some other mechanism needed here.

robert-cronin commented on pkg/ingestor/parser/common/helpers.go:
```diff
@@ -128,6 +128,9 @@ func RemoveDuplicateIdentifiers(identifierStrings *IdentifierStrings) {
 	}
+	if len(identifierStrings.OciRegistryStrings) > 0 {
+		identifierStrings.OciRegistryStrings = removeDuplicate(identifierStrings.OciRegistryStrings)
+	}
```
No use case in mind, my strategy has been to follow the OCI collector in implementing the registry collector and I think I misunderstood what identifier strings represented before. Removing this 👍

robert-cronin replied:
done

robert-cronin commented on pkg/handler/collector/oci/oci_registry.go:
```diff
@@ -0,0 +1,162 @@
+	rc := regclient.New(o.rcOpts...)
+
+	// Process each registry from the data sources
+	for _, r := range ds.OciRegistryDataSources {
```
Thank you @lumjjb Is there a link to some discussion or an issue that references the redesign of the csub interface?

robert-cronin commented:
> Awesome PR @robert-cronin , cool implementation with the recursive collectors and great testing! I have a couple comments suggesting some changes, but mostly LGTM. Can i also request we split this into the collector implementation and the CLI changes in another PR? Thanks!

I agree, this PR could do with some splitting in two. I've added the CLI-relevant changes to a new PR: #2241. This PR should be merged before that one. Thanks for the review @lumjjb ! back over to you for round 2 👍

hown3d commented:
Really nice! 
I could see having a filter for the repositories being useful. For example we have a registry with around 3000+ repos where not all of them are interesting for ingestion into GUAC.

pxp928 commented on internal/testing/mockRegistry/mock_registry.go:
```diff
@@ -0,0 +1,935 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mockregistry
```
great mocking!

pxp928 commented on pkg/handler/collector/oci/oci_registry.go:
```diff
@@ -0,0 +1,162 @@
+//
+// Copyright 2023 The GUAC Authors.
```
nit: 2024

robert-cronin replied:
oh good catch, thank you!

pxp928 reviewed (approved):
This is great! Just a very minor nit otherwise it looks great! LGTM

robert-cronin commented:
> Really nice! I could see having a filter for the repositories being useful. For example we have a registry with around 3000+ repos where not all of them are interesting for ingestion into GUAC.

I can see how that would be helpful! I'm wondering would it make sense then to have an option on the collector that takes a regex that filters out repos? e.g. `guacone collect registry example.azurecr.io --exclude ".*-deprecated$"`.

robert-cronin commented:
> This is great! Just a very minor nit otherwise it looks great! LGTM

Thank you @pxp928!
